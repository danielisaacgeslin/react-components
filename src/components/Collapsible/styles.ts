import { StyleSheet } from 'aphrodite/no-important';
import { STYLE } from '../../constants';

export default StyleSheet.create({
  bodyContainer: {
    display: 'flex',
    overflow: 'hidden',
    flexWrap: 'wrap',
    transition: 'all 0.3s',
    flex: 0
  },
  bodyContainerOpen: {
    flex: 1
  },
  header: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginBottom: 10
  },
  title: {
    textTransform: 'capitalize',
    fontSize: STYLE.FONT.SIZE.P
  },
  icon: {
    cursor: 'pointer'
  }
});
