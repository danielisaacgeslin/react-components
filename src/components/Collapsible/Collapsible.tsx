import React, { PureComponent } from 'react';
import { css } from 'aphrodite/no-important';

import Icon from '../Icon';
import styles from './styles';

export interface ICollapsibleProps {
  title: string | React.ReactElement;
  defaultOpen?: boolean;
  iconOpen?: React.ReactElement;
  iconCollapsed?: React.ReactElement;
}

export interface ICollapsibleState {
  isOpen: boolean;
  height: number;
}

export default class Collapsible extends PureComponent<ICollapsibleProps, ICollapsibleState> {
  public static defaultProps = {
    defaultOpen: true,
    iconOpen: <Icon icon={'angle-up'} />,
    iconCollapsed: <Icon icon={'angle-down'} />
  };

  public state: ICollapsibleState = {
    isOpen: false,
    height: 0
  };

  public ref = React.createRef<HTMLDivElement>();

  public componentDidMount(): void {
    const { defaultOpen } = this.props;
    this.setState({ isOpen: defaultOpen });
  }
  public componentDidUpdate(prevProps: ICollapsibleProps, prevState: ICollapsibleState): void {
    const { height } = prevState;
    setTimeout(() => {
      /* istanbul ignore if */
      if (this.ref.current && this.ref.current.scrollHeight !== height) this.setState({ height: this.ref.current.scrollHeight });
    });
  }

  public toggleCollapsible = () => this.setState(state => ({ ...state, isOpen: !state.isOpen }));

  public render() {
    const { children, title, iconOpen, iconCollapsed } = this.props;
    const { isOpen, height } = this.state;

    return (
      <div key="Collapsible">
        <header className={css(styles.header)}>
          {typeof title === 'string' ? <span className={css(styles.title)}>{title}</span> : title}
          <div data-attribute="button-toggle" className={css(styles.icon)} onClick={this.toggleCollapsible}>
            {isOpen ? iconOpen : iconCollapsed}
          </div>
        </header>
        <div ref={this.ref} className={css(styles.bodyContainer)} style={{ height: isOpen ? height : 0 }}>
          {children}
        </div>
      </div>
    );
  }
}
