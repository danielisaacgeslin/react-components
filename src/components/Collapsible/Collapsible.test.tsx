import React from 'react';
import { mount } from 'enzyme';
import { create } from 'react-test-renderer';

import Collapsible, { ICollapsibleProps } from './Collapsible';

describe('Collapsible', () => {
  let props: ICollapsibleProps;
  let wrapper;

  beforeEach(() => {
    global.console.error = () => {
      /** */
    };
    props = { title: 'Collapsible title ', defaultOpen: true };
    wrapper = mount<Collapsible>(
      <Collapsible {...props}>
        <div>Content</div>
      </Collapsible>
    );
  });

  it('should render', done => {
    setTimeout(() => {
      expect(create(wrapper).toJSON()).toMatchSnapshot();
      done();
    });
  });

  it('should render with content hidden', () => {
    props.defaultOpen = false;
    wrapper = mount<Collapsible>(
      <Collapsible {...props}>
        <div>Content</div>
      </Collapsible>
    );
    expect(create(wrapper).toJSON()).toMatchSnapshot();
  });

  it('should hide the content', () => {
    wrapper = mount<Collapsible>(
      <Collapsible {...props}>
        <div>Content</div>
      </Collapsible>
    );
    wrapper.find('[data-attribute="button-toggle"]').simulate('click');

    expect(create(wrapper).toJSON()).toMatchSnapshot();
  });

  it('should render a title as a component', () => {
    props.title = <h1>TITLE</h1>;
    wrapper = mount<Collapsible>(
      <Collapsible {...props}>
        <div>Content</div>
      </Collapsible>
    );

    expect(create(wrapper).toJSON()).toMatchSnapshot();
  });

  it('should render custom icon', () => {
    props.iconOpen = <i>icon open</i>;
    props.iconCollapsed = <i>icon collapsed</i>;
    wrapper = mount<Collapsible>(
      <Collapsible {...props}>
        <div>Content</div>
      </Collapsible>
    );

    expect(create(wrapper).toJSON()).toMatchSnapshot();
  });
});
