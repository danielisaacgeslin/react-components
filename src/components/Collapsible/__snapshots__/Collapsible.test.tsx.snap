// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible should hide the content 1`] = `
.header {
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-box-direction: normal;
  -webkit-box-orient: horizontal;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  align-items: center;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  margin-bottom: 10px;
}

.title {
  text-transform: capitalize;
  font-size: 15px;
}

.icon {
  cursor: pointer;
}

.bodyContainer {
  -webkit-box-lines: multiple;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  overflow: hidden;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-transition: all 0.3s;
  -moz-transition: all 0.3s;
  transition: all 0.3s;
  -webkit-flex: 0;
  -ms-flex: 0 1 0%;
  flex: 0;
}

<div>
  <header
    className="header"
  >
    <span
      className="title"
    >
      Collapsible title 
    </span>
    <div
      className="icon"
      data-attribute="button-toggle"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-up fa-w-10 "
        data-icon="angle-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </header>
  <div
    className="bodyContainer"
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      Content
    </div>
  </div>
</div>
`;

exports[`Collapsible should render 1`] = `
<div>
  <header
    className="header_xzo27s"
  >
    <span
      className="title_yvog1c"
    >
      Collapsible title 
    </span>
    <div
      className="icon_13xlah4"
      data-attribute="button-toggle"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-up fa-w-10 "
        data-icon="angle-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </header>
  <div
    className="bodyContainer_bbdpsy"
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      Content
    </div>
  </div>
</div>
`;

exports[`Collapsible should render a title as a component 1`] = `
.header {
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-box-direction: normal;
  -webkit-box-orient: horizontal;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  align-items: center;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  margin-bottom: 10px;
}

.icon {
  cursor: pointer;
}

.bodyContainer {
  -webkit-box-lines: multiple;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  overflow: hidden;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-transition: all 0.3s;
  -moz-transition: all 0.3s;
  transition: all 0.3s;
  -webkit-flex: 0;
  -ms-flex: 0 1 0%;
  flex: 0;
}

<div>
  <header
    className="header"
  >
    <h1>
      TITLE
    </h1>
    <div
      className="icon"
      data-attribute="button-toggle"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-up fa-w-10 "
        data-icon="angle-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </header>
  <div
    className="bodyContainer"
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      Content
    </div>
  </div>
</div>
`;

exports[`Collapsible should render custom icon 1`] = `
.header {
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-box-direction: normal;
  -webkit-box-orient: horizontal;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  align-items: center;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  margin-bottom: 10px;
}

.title {
  text-transform: capitalize;
  font-size: 15px;
}

.icon {
  cursor: pointer;
}

.bodyContainer {
  -webkit-box-lines: multiple;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  overflow: hidden;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-transition: all 0.3s;
  -moz-transition: all 0.3s;
  transition: all 0.3s;
  -webkit-flex: 0;
  -ms-flex: 0 1 0%;
  flex: 0;
}

<div>
  <header
    className="header"
  >
    <span
      className="title"
    >
      Collapsible title 
    </span>
    <div
      className="icon"
      data-attribute="button-toggle"
      onClick={[Function]}
    >
      <i>
        icon open
      </i>
    </div>
  </header>
  <div
    className="bodyContainer"
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      Content
    </div>
  </div>
</div>
`;

exports[`Collapsible should render with content hidden 1`] = `
.header {
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-box-direction: normal;
  -webkit-box-orient: horizontal;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  align-items: center;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  margin-bottom: 10px;
}

.title {
  text-transform: capitalize;
  font-size: 15px;
}

.icon {
  cursor: pointer;
}

.bodyContainer {
  -webkit-box-lines: multiple;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  overflow: hidden;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-transition: all 0.3s;
  -moz-transition: all 0.3s;
  transition: all 0.3s;
  -webkit-flex: 0;
  -ms-flex: 0 1 0%;
  flex: 0;
}

<div>
  <header
    className="header"
  >
    <span
      className="title"
    >
      Collapsible title 
    </span>
    <div
      className="icon"
      data-attribute="button-toggle"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-down fa-w-10 "
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </header>
  <div
    className="bodyContainer"
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      Content
    </div>
  </div>
</div>
`;
